cmd/registry/controller/controller.go

```diff
@@ -232,73 +233,64 @@ func generateCreateActions(
-	// 1. Target resource is a project level artifact "projects/demo/locations/global/artifacts/serach-index"
-	//    extracted parent will be "projects/demo/locations/global"
-	// 1. Target resource is an api "projects/demo/locations/global/apis/petstore"
-	//    extracted parent will be "projects/demo/locations/global"
-	if strings.HasSuffix(parentName, "locations/global") {
+
+	parentName := parsedResourcePattern.ParentName()
+	switch parentName.(type) {
+	case projectName:
+		// If parent is a project, we can't list projects since this is registry client command.
+		// Since the manifest definition is scoped  only for a particular project,
+		// there will be only one target resource in this case.
+		// There are two cases where this might happen:
+		// 1. Target resource is a project level artifact "projects/demo/locations/global/artifact/search-index"
+		//    extracted parent will be "projects/demo/locations/global"
+		// 1. Target resource is an api "projects/demo/locations/global/apis/petstore"
+		//    extracted parent will be "projects/demo/locations/global"
 		// Return if this parent was already visited.
-		if visited[parentName] {
+		if visited[parentName.String()] {
 			return nil, nil
 		}
-
-		// Since the GeneratedResource is non-existent here,
-		// we will have to derive the exact name of the target resource.
-		targetResourceName, err := resourceNameFromParent(resourcePattern, parentName)
-		if err != nil {
-			return nil, fmt.Errorf("Cannot generate target resourceName to be created. Error: %s", err)
+		parentList = []resourceInstance{
+			projectResource{
+				projectName: parentName,
+			},
 		}
 
-		takeAction, err := needsCreate(
-			targetResourceName,
-			dependencyMaps,
-			generatedResource,
-		)
-
+	default:
+		// If parent resource is not a project, then go through all the non-visited parents.
+		visitedJSON, err := json.Marshal(visited)
 		if err != nil {
-			return nil, err
-		} else if !takeAction {
-			return nil, nil
+			return nil, fmt.Errorf("internal error: Invalid visited map, %s", err)
 		}
-
-		cmd, err := generateCommand(generatedResource.Action, targetResourceName.String())
+		// Construct a filter to fetch only non-visited parents.
+		// Check that the "name" of the resource is not in the visited list of parents.
+		// Example:
+		// Consider a visited map of parents which are apis:
+		// Visited: {
+		//	"projects/demo/locations/global/apis/example-api1": true,
+		//  "projects/demo/locations/global/apis/example-api2": true,
+		// }
+		// We make use of `e.all(x,p)` macro as defined here: https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros
+		// The CEL filter for listing non-visited parents will be as follows:
+		// '{
+		//	"projects/demo/locations/global/apis/example-api1": true,
+		//  "projects/demo/locations/global/apis/example-api2": true,
+		// }.all(n, !(name.contains(n)))'
+		filter := fmt.Sprintf("%s.all(n, !(name.contains(n)))", visitedJSON)
```

Comment:
I think the explanation is clear. Do we use `contains` instead of `==` because we want to ignore all resources whose parent has been visited? Should we be using `startsWith` instead?

It doesn't seem like we need to use `json.Marshal` here either. That was one of the things that confused me when I first looked at the code because I wasn't sure how a serialized map would work with the `all` macro. Building the JSON array using string processing can avoid the possibility of an error, as a bonus.

This can be captured in a separate function to help make the code a bit more self-documenting and reduce the noise in this func.

```
// Constructs a CEL filter to exclude resources with visited parents.
// The filter excludes resources whose `name` property starts with any of the visited parent names.
// 
// For example, consider a visited map of parents which are apis:
// {
//     "projects/demo/locations/global/apis/example-api1": true,
//     "projects/demo/locations/global/apis/example-api2": true,
// }
//
// The resulting CEL filter will be:
// ["projects/demo/locations/global/apis/example-api1","projects/demo/locations/global/apis/example-api2"].all(prefix, !name.startsWith(prefix))
//
// Note: The `bool` values in the input map are ignored. The filter will use every map key. 
func excludeVisitedParents(v map[string]bool) string {
	// Wrap each string with quotes and join them with commas to build a JSON string array.
	jsonStrings := make([]string, 0, len(v))
	for parent := range v {
		jsonStrings = append(jsonStrings, fmt.Sprintf("%q", parent))
	}
	return fmt.Sprintf("[%s].all(parent, !name.startsWith(parent))", strings.Join(jsonStrings, ","))
}
```

Reply:
Thanks for this suggestion.

Created a separate function as suggested. 
For your question on the use of `contains`, yes, the original intention was to ignore all the resources with parents as prefixes. However, after doing another pass at it, I realized that there is no case right now where, the `List` fetches children resources of the parents which are stored in the visited map.  The expectation is that the values stored in the visited map should be full resource names and not prefixes to any of the parents. Use of `contains` opens the possibility for bugs in the logic which might go unnoticed.

I am going to revert this to ==` operation for now, we can always change it to `startsWith` if we feel the need for it in the future.